### DISCUSSION TITLE
Mapbox Overlay blend transparent PNG

### DISCUSSION THREAD

**Discussion by MartinP-C** [visgl/deck.gl#7427] [Q&A] [answered]
I am migrating some layers from Mapbox to using Deck.gl in Mapbox Overlay mode.

Originally the tiled images are bright and colourful as in this example which is Mapbox only:
![image](https://user-images.githubusercontent.com/8778016/202719090-57e4026f-ee11-436c-aa7b-7926cccf8304.png)

When I put the same image using BitmapLayer it is noticeably darker and more dull:
![image](https://user-images.githubusercontent.com/8778016/202719518-e0ea23ad-4a4c-4665-bb7b-c178035eae53.png)

I have tried `transparentColor: [255, 255, 255, 0]` in `parameters`, it certainly brightens the images but the color is still washed out:
![image](https://user-images.githubusercontent.com/8778016/202719720-ebe638d2-28d2-4b03-aef0-880cae7719e3.png)

I have tried all the values for CSS `mix-blend-mode` but none of those achieve the same effect.
I have also tried lots of combinations of WebGL parameters for blend functions but none get close to the original, although I am no WebGL expert at all.

I suspect it is maybe impossible to blend my overlay with the map in the same way as mapbox only because they are separate canvases, however, can I achieve something close to it?

**Comment by chrisgervang** [comment]
Have you tried [interleaving](https://deck.gl/docs/api-reference/mapbox/mapbox-overlay#interleaved) the overlay? That'll render deck into the same webgl context as mapbox, and is also how you'd put the bitmap layer under the text layers as you did before using the `beforeId` prop. If you run into transparency issues while interleaving, try adding these parameters:
https://github.com/visgl/deck.gl/blob/1342a961ba9ce320ddd0eebe1cd53a021c3c8382/modules/mapbox/src/deck-utils.ts#L46-L54

**Reply by MartinP-C** [reply]
Thanks. I gave that a go and it didn't make any difference to the colours.
But perhaps it is because the layer interleaving is not working for me?
with `interleaved: true` I now have a single canvas, but the Deck layers are not beneath the label layers.

**Comment by Pessimistress** [comment]
This is a legacy issue in the BitmapLayer. If you set `transparencyColor` to `[255, 0, 0, 0]` in the above example you will get the expected color. I will put up a fix for it, but it will unlikely make it into a 8.8.x patch since it's a breaking change.

**Reply by MartinP-C** [reply]
Thanks for looking at this. With red as the transparency I just get red-tinted images:
![image](https://user-images.githubusercontent.com/8778016/203278524-7ec00e35-0b08-4b64-84b9-e5d66d36a9e7.png)

I have changed the test image so here it is in mapbox:
![image](https://user-images.githubusercontent.com/8778016/203278578-a0f4e84e-0c7e-485e-ac8a-e4cf7f5fe80b.png)

I am using the BitmapLayer inside a TileLayer, which is also inside a custom layer class.

Overlay:
```tsx
const DeckOverlay: FC = () => {
  const layers = [ new DeckTilePoly() ];

  const overlayProps: OverlayProps = { layers, interleaved: true };
  const mapboxOverlay = new MapboxOverlay(overlayProps);

  const overlay = useControl<MapboxOverlay>(() => {
    return mapboxOverlay;
  });
  overlay.setProps(overlayProps);

  return null;
};

export default DeckOverlay;
```

custom layer:
```ts

class DeckTilePoly extends CompositeLayer {
  static layerName = 'DeckTilePoly';
  swapLevel = 15;

  shouldUpdateState({ changeFlags }: any) {
    // console.log('changeFlags', changeFlags);
    return changeFlags.somethingChanged;
  }

  renderLayers() {
    return [
      new TileLayer({
        visible: this.context.viewport.zoom < this.swapLevel,
        id: 'tile-images',
        minZoom: 0,
        maxZoom: this.swapLevel + 1,
        beforeId: 'waterway-label',
        zoomOffset: 1,
        getTileData: async (props) => {
          const { x, y, z } = props.index;
          const { signal } = props;
          const response = await axios.get<ArrayBuffer>(`/${z}/${x}/${y}`, {
            responseType: 'arraybuffer',
            signal,
          });

          if (response.status === 200) {
            const data = response.data;

            const a = await load(data, ImageLoader);

            return a;
          }
          return null;
        },

        renderSubLayers: (props) => {
          const bbox = props.tile.bbox as GeoBoundingBox;
          const { west, south, east, north } = bbox;

          return new BitmapLayer(props, {
            data: null,
            image: '/test.png',
            // image: props.data,
            bounds: [west, south, east, north],
            transparentColor: [255, 0, 0, 0],
            parameters: {
              depthMask: true,
              depthTest: true,
              blend: true,
              blendFunc: [gl.SRC_ALPHA, gl.ONE_MINUS_SRC_ALPHA, gl.ONE, gl.ONE_MINUS_SRC_ALPHA],
              polygonOffsetFill: true,
              depthFunc: gl.LEQUAL,
              blendEquation: gl.FUNC_ADD,
            },
          });
        },
      }),
      new TileLayer({
        visible: this.context.viewport.zoom >= this.swapLevel,
        id: 'tile-polys',
        minZoom: this.swapLevel,
        getTileData: async (props) => {
          const { x, y, z } = props.index;
          const { west, south, east, north } = props.bbox as GeoBoundingBox;
          const { signal } = props;

          const data = await axios.get(`/results`, {
            params: {
              x,
              y,
              z,
            },
            signal,
          });

          return data;
        },
        renderSubLayers: (props) => {
          return new GeoJsonLayer(props, {
             // set colors etc.
          });
        },
      }),
    ];
  }
}

export default DeckTilePoly;
```

**Reply by Pessimistress** [reply]
There is more explanation on this bug in #7441.

Given that it's a breaking change, the PR is not going to make it into a patch. As a workaround, you can try add this to the BitmapLayer:

```js
parameters: {
  blendFunc: [1, 1]
}
```

**Reply by MartinP-C** [reply]
It seems applying `blendFunc: [1, 1]` gives even more of a pink tint:
![image](https://user-images.githubusercontent.com/8778016/206141480-da0b9bf5-46e9-4466-9243-a08dbcae8c5d.png)

Is there any timeline on when a version with the PR might be released? Appreciate however that's likely a big release and not likely to be soon.
Just to inform whether I wait for an official release, maybe try a forked build, keep trying working around with WebGL, or rethink the images entirely :)

**Reply by Pessimistress** [reply]
Please unset any tintColor when you use the work around.

**Reply by MartinP-C** [reply]
That screenshot is with no tintColor set.
The settings for the latest image are:
```ts
        renderSubLayers: (props) => {
          const bbox = props.tile.bbox;
          const { west, south, east, north } = bbox;

          return new BitmapLayer(props, {
            data: null,
            image: '/test.png',
            bounds: [west, south, east, north],
            parameters: {
              depthMask: true,
              depthTest: true,
              blend: true,
              blendFunc: [1, 1],
              polygonOffsetFill: true,
              depthFunc: gl.LEQUAL,
              blendEquation: gl.FUNC_ADD,
            },
            transparentColor: [255, 0, 0, 0],
          });
        },
```

**Reply by MartinP-C** [reply]
I've created a codesandbox with that test image added: https://codesandbox.io/s/reverent-framework-in56yf?file=/src/index.js

**Reply by Pessimistress** [reply]
Unset transparentColor as well.

**Reply by MartinP-C** [reply]
Huge apologies, I completely overlooked that I'm using Deck in an overlay.
I've updated the sandbox

**Reply by Pessimistress** [reply]
Alright try this

```js
import GL from '@luma.gl/constants'

new BitmapLayer({
  ...
  parameters: {
     blendFunc: [GL.ONE, GL.ONE_MINUS_SRC_ALPHA]
  }
})
```

**Reply by MartinP-C** [reply, marked as answer]
It seems that I've found a solution.
In combination with the suggestion to use `blendFunc: [GL.ONE, GL.ONE_MINUS_SRC_ALPHA]`, I disabled the `premultiplyAlpha` option in the ImageLoader:
```js
await load(data, ImageLoader, {
            imagebitmap: { premultiplyAlpha: "none" },
            image: { type: "imagebitmap" },
          });
```

Neither change works by itself, but together it makes my transparent image overlay with the exact colours I am expecting.
